### PR TITLE
Fad api changes

### PR DESCRIPTION
**Pull request by mannyjimenez0810**
FAD Adding in changes asked by the API council.

### CODE REVIEW

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 1]
```diff
@@ -111,22 +123,53 @@ public static FirebaseAppDistribution getInstance(@NonNull FirebaseApp app) {
...
+    synchronized (updateTaskLock) {
+      if (cachedUpdateToLatestReleaseTask != null
+          && !cachedUpdateToLatestReleaseTask.isComplete()) {
+        return cachedUpdateToLatestReleaseTask;
+      }
+
+      cachedUpdateToLatestReleaseTask = new UpdateTaskImpl();
+    }
+    checkForNewRelease()
+        .onSuccessTask(
```
Comment (pranavrajgopal) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:136]: Any particular reason we are using an executor here? If we do need an executor, we should create one for the main thread. I don't see why we need to implement this on a separate thread.

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 2]
```diff
@@ -296,27 +338,61 @@ AppDistributionReleaseInternal getCachedLatestRelease() {
...
     alertDialog.setButton(
         AlertDialog.BUTTON_NEGATIVE,
         context.getString(R.string.update_no_button),
         (dialogInterface, i) -> {
           dialogInterface.dismiss();
-          updateAlertDialogTask.setException(
-              new FirebaseAppDistributionException(
-                  Constants.ErrorMessages.UPDATE_CANCELED,
-                  FirebaseAppDistributionException.Status.INSTALLATION_CANCELED));
+          synchronized (updateTaskLock) {
```
Comment (pranavrajgopal) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:357]: I think it might be worth extracting the updateProgress setting and the `setException` into a method since they are being called in a couple of places. Added advantage - will reduce the number of synchronization blocks.

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 3]
```diff
@@ -296,27 +338,61 @@ AppDistributionReleaseInternal getCachedLatestRelease() {
...
-    return updateAlertDialogTask.getTask();
+    synchronized (updateTaskLock) {
+      return cachedUpdateToLatestReleaseTask;
+    }
   }
 
   void setInstallationResult(int resultCode) {
     this.updateAppClient.setInstallationResult(resultCode);
   }
+
```
Comment (pranavrajgopal) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:380]: might be worth renaming this to be more specific for the updateToLatestRelease task

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 4]
```diff
@@ -111,22 +123,53 @@ public static FirebaseAppDistribution getInstance(@NonNull FirebaseApp app) {
...
   @NonNull
-  public synchronized Task<Void> updateToLatestRelease() {
-    if (cachedUpdateToLatestReleaseTask != null && !cachedUpdateToLatestReleaseTask.isComplete()) {
+  public synchronized UpdateTask updateIfNewReleaseAvailable() {
+    synchronized (updateTaskLock) {
+      if (cachedUpdateToLatestReleaseTask != null
+          && !cachedUpdateToLatestReleaseTask.isComplete()) {
+        return cachedUpdateToLatestReleaseTask;
+      }
+
```
Comment (pranavrajgopal) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:132]: I think we should rename this task `cachedUpdateIfNewReleaseAvailable` to reflect the change in the method name

**Review by pranavrajgopal — APPROVED** [review]

**Comment by pranavrajgopal** [comment]
Nice work!!

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java (hunk 4 above)
Comment (mannyjimenez0810) [reply firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:132]: Yea most name changes will happen in an additional PR.

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java (hunk 2 above)
Comment (mannyjimenez0810) [reply firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:357]: The updateProgress with a setException in the same block only happens once. the other time the pieces of code are close they are in a failure listen and on progress listener. So I am not sure it makes much sense to take it out since it seems this combo only happens once. In other places there are either a setTaskCompletion or a setResult.

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java (hunk 3 above)
Comment (mannyjimenez0810) [reply firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:380]: Agreed. Naming changes are planned to go into a separate PR as of now.

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 5]
```diff
@@ -141,16 +184,16 @@ public static FirebaseAppDistribution getInstance(@NonNull FirebaseApp app) {
...
+  public synchronized Task<AppDistributionRelease> checkForNewRelease() {
     if (cachedCheckForUpdateTask != null && !cachedCheckForUpdateTask.isComplete()) {
       LogWrapper.getInstance().v("Response in progress");
       return cachedCheckForUpdateTask;
     }
-
     cachedCheckForUpdateTask =
         signInTester()
-            .onSuccessTask(unused -> this.checkForUpdateClient.checkForUpdate())
+            .onSuccessTask(newReleaseExecutor, unused -> this.checkForUpdateClient.checkForUpdate())
```
Comment (rachaprince) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:194]: we may want to try removing these to see if the UI thread can handle this as well

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 6]
```diff
@@ -141,16 +184,16 @@ public static FirebaseAppDistribution getInstance(@NonNull FirebaseApp app) {
...
             .onSuccessTask(
+                newReleaseExecutor,
                 appDistributionReleaseInternal -> {
                   setCachedLatestRelease(appDistributionReleaseInternal);
                   return Tasks.forResult(
```
Comment (rachaprince) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:199]: ditto

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 7]
```diff
@@ -141,16 +184,16 @@ public static FirebaseAppDistribution getInstance(@NonNull FirebaseApp app) {
...
       LogWrapper.getInstance().v("Response in progress");
       return cachedCheckForUpdateTask;
     }
-
     cachedCheckForUpdateTask =
         signInTester()
-            .onSuccessTask(unused -> this.checkForUpdateClient.checkForUpdate())
+            .onSuccessTask(newReleaseExecutor, unused -> this.checkForUpdateClient.checkForUpdate())
             .onSuccessTask(
+                newReleaseExecutor,
```
Comment (rachaprince) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:196]: ditto

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 8]
```diff
@@ -296,27 +324,71 @@ AppDistributionReleaseInternal getCachedLatestRelease() {
...
+  private void setCachedUpdateIfNewReleaseCompletionError(
+      Exception e, FirebaseAppDistributionException defaultFirebaseException) {
+    if (e instanceof FirebaseAppDistributionException) {
+      setCachedUpdateIfNewReleaseCompletionError((FirebaseAppDistributionException) e);
+    } else {
+      setCachedUpdateIfNewReleaseCompletionError(defaultFirebaseException);
+    }
+  }
+
+  private void cachedUpdateIfNewReleaseUpdateProgress(UpdateProgress progress) {
```
Comment (rachaprince) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:383]: nit: Java methods often start with verbs. Maybe this could be something like `postProgressToCachedUpdateIfNewReleaseTask` ?

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 9]
```diff
@@ -296,27 +324,71 @@ AppDistributionReleaseInternal getCachedLatestRelease() {
...
+    synchronized (updateTaskLock) {
+      cachedUpdateIfNewReleaseTask.updateProgress(progress);
+    }
+  }
+
+  private void cachedUpdateIfNewReleaseTaskSetResult() {
```
Comment (rachaprince) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:389]: similar to the above, maybe rephrase as `setCachedUpdateIfNewReleaseResult` to align with naming conventions?

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java [hunk 10]
```diff
@@ -296,27 +324,71 @@ AppDistributionReleaseInternal getCachedLatestRelease() {
         AlertDialog.BUTTON_POSITIVE,
         context.getString(R.string.update_yes_button),
         (dialogInterface, i) ->
-            // show download progress in notification manager
+        // show download progress in notification manager
```
Comment (rachaprince) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:327]: nit: this comment is explaining why updateApp is called with `true`, so we should probably move it to right above line 330

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistributionNotificationsManager.java [hunk 11]
```diff
@@ -61,10 +61,14 @@ void updateNotification(long totalBytes, long downloadedBytes, UpdateStatus stat
     notificationManager.notify(NOTIFICATION_TAG, /*id =*/ 0, builder.build());
   }
 
+  // CHECK THIS LATER
   private boolean isErrorState(UpdateStatus status) {
     return status.equals(UpdateStatus.DOWNLOAD_FAILED)
         || status.equals(UpdateStatus.INSTALL_FAILED)
-        || status.equals(UpdateStatus.INSTALL_CANCELED);
+        || status.equals(UpdateStatus.INSTALL_CANCELED)
+        || status.equals(UpdateStatus.NEW_RELEASE_NOT_AVAILABLE)
```
Comment (rachaprince) [inline firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistributionNotificationsManager.java:69]: `NEW_RELEASE_NOT_AVAILABLE` isn't an error state, so we can probably remove it.

File: firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/FirebaseAppDistributionTest.java [hunk 12]
```diff
@@ -236,6 +244,7 @@ public void updateApp_whenNotSignedIn_throwsError() throws Exception {
   @Test
   public void updateToLatestRelease_whenNewAabReleaseAvailable_showsUpdateDialog()
       throws Exception {
+    shadowOf(Looper.getMainLooper()).idle();
```
Comment (rachaprince) [inline firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/FirebaseAppDistributionTest.java:247]: Did we figure out if we need this? Or did adding the `addcancelationListener` fix the issue this was trying to solve

File: firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/FirebaseAppDistributionTest.java [hunk 13]
```diff
@@ -268,14 +277,15 @@ public void updateToLatestRelease_whenNewAabReleaseAvailable_showsUpdateDialog()
   @Test
   public void updateToLatestRelease_whenReleaseNotesEmpty_doesNotShowReleaseNotes()
       throws Exception {
+    shadowOf(Looper.getMainLooper()).idle();
```
Comment (rachaprince) [inline firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/FirebaseAppDistributionTest.java:280]: same question as above

File: firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/FirebaseAppDistributionTest.java [hunk 14]
```diff
@@ -397,4 +421,42 @@ public void updateApp_appResume_tryResetAabUpdateTask() {
 
     verify(mockUpdateAppClient, times(1)).tryCancelAabUpdateTask();
   }
+
+  @Test
+  public void updateToLatestRelease_recieveProgressUpdateFromUpdateApp() throws Exception {
+    shadowOf(Looper.getMainLooper()).idle();
+    // mockSignInStorage returns false then true to simulate logging in during first signIn check in
```
Comment (rachaprince) [inline firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/FirebaseAppDistributionTest.java:428]: nit: remove comment

**Review by rachaprince — COMMENTED** [review]
Nice work!

It looks like check-changed and device-check-changed are failing. Clicking on the details tab, I noticed:

```
 Did you mean '@Override @NonNull'?
warning: [MissingOverride] addOnCanceledListener overrides method in Task; expected @Override
[I] 2021-09-03 15:25:53 +0000 UTC fireci.gradle:   public Task<Void> addOnCanceledListener(
```

maybe adding a @NonNull annotation will fix the failing tests?
 (If you want to debug the test locally you can run `./gradlew [name of test that failed]`)

File: firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/FirebaseAppDistributionTest.java (hunk 12 above)
Comment (mannyjimenez0810) [reply firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/FirebaseAppDistributionTest.java:242]: We did not need this. removing now.

File: firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java (hunk 1 above)
Comment (mannyjimenez0810) [reply firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java:128]: removed executor

**Review by rachaprince — APPROVED** [review]
